scripts/validate_exception_location.py

```diff
@@ -0,0 +1,123 @@
+"""
+Validate that the exceptions and warnings are in approrirate places.
+
+Checks for classes that inherit a python exception and warning and
+flags them, unless they are exempted from checking.
+
+Print the exception/warning that do not follow convention.
+
+Usage::
+
+As pre-commit hook (recommended):
+    pre-commit run validate-errors-locations --all-files
+"""
+from __future__ import annotations
+
+import argparse
+import ast
+import sys
+from typing import Sequence
+
+ERROR_MESSAGE = (
+    "{path}:{lineno}:{col_offset}: {exception_name}: "
+    "Please don't place exceptions or warnings outside of pandas/errors/__init__.py or "
+    "pandas/_libs\n"
+)
+exception_warning_list = {
+    "ArithmeticError",
+    "AssertionError",
+    "AttributeError",
+    "EOFError",
+    "Exception",
+    "FloatingPointError",
+    "GeneratorExit",
+    "ImportError",
+    "IndentationError",
+    "IndexError",
+    "KeyboardInterrupt",
+    "KeyError",
+    "LookupError",
+    "MemoryError",
+    "NameError",
+    "NotImplementedError",
+    "OSError",
+    "OverflowError",
+    "ReferenceError",
+    "RuntimeError",
+    "StopIteration",
+    "SyntaxError",
+    "SystemError",
+    "SystemExit",
+    "TabError",
+    "TypeError",
+    "UnboundLocalError",
+    "UnicodeDecodeError",
+    "UnicodeEncodeError",
+    "UnicodeError",
+    "UnicodeTranslateError",
+    "ValueError",
+    "ZeroDivisionError",
+    "BytesWarning",
+    "DeprecationWarning",
+    "FutureWarning",
+    "ImportWarning",
+    "PendingDeprecationWarning",
+    "ResourceWarning",
+    "RuntimeWarning",
+    "SyntaxWarning",
+    "UnicodeWarning",
+    "UserWarning",
+    "Warning",
+}
+
+permisable_exception_warning_list = [
+    "LossySetitemError",
+    "NoBufferPresent",
+    "InvalidComparison",
+    "NotThisMethod",
+    "OptionError",
+    "InvalidVersion",
+]
+
+
+class Visitor(ast.NodeVisitor):
+    def __init__(self, path: str) -> None:
+        self.path = path
+
+    def visit_ClassDef(self, node):
+        classes = {getattr(n, "id", None) for n in node.bases}
+
+        if (
+            classes
+            and classes.issubset(exception_warning_list)
```

Comment:
Not great we have to keep another list of pandas defined warnings. Can we check if the defined Warning/Exception came from `pandas/`?

Reply:
Sorry for sleeping on this for so long. I'm not sure if I understand the question.

It's good if we can remove `permisable_exception_warning_list`, correct? I don't like it either, but I not sure how to handle preexisting exception outside of `errors/__init__.py` and `_libs`. For example, I could tell `LossySetitemError` came from `pandas/core/dtypes/cast.py`. But, if someone added another exception to `cast.py` then I'm not sure how to flag the new exception, but allow the old exception. 

Does that get at the question or am I off base?

Reply:
> It's good if we can remove permisable_exception_warning_list, correct?

Yeah this is my main concern. IMO any pandas error (whether just internal or exposed externally) should be imported into `pandas/errors/__init__.py`

Maybe one approach would be

1. Grab all the warnings/exceptions from `doc/source/reference/testing.rst` (this file should be in sync with `errors/__init__.py` through another check
2. AST parse through `pandas/` (excluding tests) to record all defined warning/exception subclasses
3. Compare the results.

Reply:
Gotcha, thank you for the help and the clarification. I've updated the MR to leverage `testing.rst`. Let me know if this solution works better

Reply:
This line can be removed now right? The Warning/Exception name should be just checked against `self.exception_set` now?

Reply:
I think I still need that line. I sort of kept my original approach. This approach was to have `Visitor` class check if the class was a warning/exception subclass and check if the class was in the testing.rst. Now, I've reworked to be more inline with your approach. So, `Visitor` gets all the subclass exceptions and then `validate_exception_and_warning_placement` will use `self.exception_set` to find the difference. 

Let me know if that is reasonable/make sense

Reply:
Yeah, I don't really like the giant list either. AFAIK, the base is a `ast.Name` object, and that really only has an `id` which acts as the name: https://docs.python.org/3/library/ast.html#ast.Name. 

However, if we're comfortable assuming that if a base name is `Exception` or if the name ends with either `Error` or `Warning` then that's an `Exception` or  `Warning`.  I think we could re-write it like the below. It's slightly less hard-codey. Let me know if that shows good/reasonable. 

```
    def visit_ClassDef(self, node) -> None:
        def is_an_exception_subclass(base_id: str) -> bool:
            return base_id == "Exception" or base_id.endswith("Warning") or base_id.endswith("Error")

        classes = []
        for n in node.bases:
            base_id = getattr(n, "id", None)
            if base_id and is_an_exception_subclass(base_id):
                classes.append(base_id)

        if classes:
            self.possible_exceptions.add(node.name)
```

Reply:
Hmm I tried something like this and was able to get the name of the base class

```
In [1]: import ast

In [2]: text = """
   ...: class MyWarning(Warning):
   ...:     pass
   ...:
   ...: class MyClass:
   ...:     pass
   ...: """

In [12]: class Visit(ast.NodeVisitor):
    ...:     def visit_ClassDef(self, node):
    ...:         breakpoint()
    ...:         return node
    ...:

In [13]: p = ast.parse(text)

In [14]: Visit().visit(p)
> <ipython-input-12-85c0ebda4faf>(4)visit_ClassDef()
-> return node
(Pdb) node
<ast.ClassDef object at 0x125468760>
(Pdb) node.bases
[<ast.Name object at 0x125468640>]
(Pdb) node.bases[0].id
'Warning'

```

Reply:
I could be off-base, but sounds like we are on the same page. I'm grabbing the ids from the bases and comparing the values. I've update the MR as well. Let me know if I'm missing something

Reply:
Ah gotcha. Yeah from the newer diff this is what I had in mind. Thanks!